FlowCrypt/Functionality/Mail Provider/Imap/Imap+retry.swift

```diff
@@ -7,117 +7,118 @@
-                lastErr[op] = error
-
-                renewSession()
-                    .then { _ in
-//                        self?.logger.logInfo("forced session refreshes \(op)")
-                        retry().then(resolve).catch(reject)
-                    }
-                    .catch(reject)
-//                logger.logDebug("just set lastErr to \(lastErr[op])")
-
-                return false
-            case .connection:
-                if let operation = lastErr[op], operation == error { return true }
-//                logDebug(13, "(\(debugId)|\(op)) it's a retriable conn err, clear sessions")
-
-                imapSess = nil // the connection has dropped, so it's probably ok to not officially "close" it
-                smtpSess = nil // but maybe there could be a cleaner way to dispose of the connection?
-                lastErr[op] = error
-
-//                logDebug(14, "(\(debugId)|\(op)) just set lastErr to ", value: lastErr[op])
-//                log("conn drop for \(op), cleared sessions, will retry \(op)", error: nil, res: nil, start: start)
-                retry().then(resolve).catch(reject)
-//                logDebug(15, "(\(debugId)|\(op)) return=true (need to retry)")
-                return false
-            default:
-//                logDebug(8, "(\(debugId)|\(op)) err not retriable, rejecting ", value: err)
-                reject(error)
-                lastErr[op] = error
-//                logDebug(9, "(\(debugId)|\(op)) just set lastErr to ", value: lastErr[op])
-//                logDebug(12, "(\(debugId)|\(op)) return=true (no need to retry)")
-                return true
+    private func asAsync(
+        _ imapSess: MCOIMAPSession,
+        _ executor: @escaping (MCOIMAPSession, @escaping (Error?) -> Void) -> Void
+    ) async throws {
+        return try await withCheckedThrowingContinuation { continuation in
+            executor(imapSess) { error in
+                if let error = error {
+                    return continuation.resume(throwing: error)
+                } else {
+                    return continuation.resume()
+                }
             }
-        } else {
-            lastErr.removeValue(forKey: op)
+        }
+    }
+
+    func shouldRetryOnce(
+        _ op: String,
+        _ appErr: AppErr
+    ) async throws -> Bool {
+        switch appErr {
+        case .authentication:
+            try await renewSession() // todo - log time
+            return true
+        case .connection:
+            // the connection has dropped, so it's probably ok to not officially "close" it. but maybe there could be a cleaner way to dispose of the connection?
+            imapSess = nil
+            smtpSess = nil
+            // this is a mess, neads a real refactor. use DI
+            // todo - for now renewing of session disabled, will probably break retries
+//            setupSession()
+//            await connectImap(session: imapSessNotNil)
             return true
```

Comment:
Todo - this is likely completely broken and will need a fix. But not in this PR.

Also, `setupSession` and `connectImap` and `connectSmpt` all need complete refactoring.